compiler/src/lnntoamm/Statement.ts

```diff
@@ -0,0 +1,785 @@
+      this.__ammName = amm.global('const', this.val.ty.breakdown(), this.val.val);
+    } else {
+      throw new Error(`unexpected expression: ${this.val}`);
+    }
+  }
+}
+
+export class FnArg extends Dec {
+  get ammName(): string {
+    if (super.val != null) {
+      if (!(super.val instanceof Ref)) {
+        throw new Error(`expected fn arg to be set to a reference (here it is: ${super.val})`);
+      }
+      return super.val.ammName;
+    } else {
+      return this.name;
+    }
+  }
+
+  set ammName(to: string) {
+    if (super.val === null) {
+      throw new Error(`cannot rename variables that aren't being inlined`);
+    } else {
+      super.ammName = to;
+    }
+  }
+
+  constructor(
+    ast: LPNode,
+    name: string,
+    ty: Type,
+  ) {
+    super(ast, true, name, ty, null);
+    if (ty === null) {
+      TODO('function params without a type specified');
+    }
+  }
+
+  static fromArgAst(ast: LPNode, metadata: MetaData): FnArg {
+    let name = ast.get('variable').t;
+    let typename = ast.get('fulltypename');
+    let argTy = Type.getFromTypename(typename, metadata.scope);
+    if (argTy === null) {
+      TODO('args with implicit types');
+    } else if (!(argTy instanceof Type)) {
+      throw new Error(`Function argument is not a valid type: ${typename.t}`);
+    }
+    const arg = new FnArg(ast, name, argTy);
+    metadata.define(arg);
+    const dec = metadata.var(arg.name);
+    if (dec !== arg) {
+      throw new Error('ugggghhhhh');
+    }
+    return arg;
+  }
+
+  ammOut(): [string, Builtin] {
+    return TODO('TODO:')
+  }
+}
+
+export class Emit extends Stmt {
+  event: Event
+  emitVal: Ref | null
+
+  constructor(
+    ast: LPNode,
+    event: Event,
+    emitVal: Ref | null,
+  ) {
+    super(ast);
+    this.event = event;
+    this.emitVal = emitVal;
+  }
+
+  static fromAst(ast: LPNode, metadata: MetaData): Stmt[] {
+    let stmts: Stmt[] = [];
+    if (ast.get('retval').has()) {
+      let emitVal = ast.get('retval').get('assignables');
+      let emitValSplit = Stmt.fromAssignables(emitVal, metadata);
+      stmts.push(...emitValSplit);
+    }
+    const eventName = ast.get('eventname').t;
+    const event = metadata.scope.deepGet(eventName);
+    if (event === null) {
+      throw new Error(`event ${eventName} not defined`);
+    } else if (!(event instanceof Event)) {
+      throw new Error(`cannot emit to non-events (${eventName} is not an event)`);
+    }
+    const emitVal = stmts[stmts.length - 1];
+    if (!(emitVal instanceof Dec) && !(emitVal instanceof Ref)) {
+      throw new Error('no declaration or reference created for emit value');
+    }
+    const emitRef = emitVal.ref();
+    const emit = new Emit(ast, event, emitRef)
+    emit.emitVal.dec.ty.constrain(event.eventTy);
+    stmts.push(emit);
+    return stmts;
+  }
+
+  pushAMM(indent: string, output: string) {
+    output.concat(
+      indent,
+      'emit ',
+      this.event.ammName,
+      ' ',
+      this.emitVal.ammName,
```

Comment:
Maybe only include these two if the event has a value to emit and isn't just a `void` event?